### PR TITLE
[Enhancement] Speed up Registry initialization

### PR DESCRIPTION
**Pull request by astahlman**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

This PR addresses https://github.com/open-mmlab/mmcv/issues/1843 by improving the performance of
`Registry.__init__()`

## Modification

Instead of calling inspect.stack() to read the entire stack and its
associated source files from disk, walk up the stack to get only the
specific frame that we need (see [this answer on StackOverflow](https://stackoverflow.com/a/42636264/895769)
for additional information).

This change results in a ~2.5x speed-up when importing from downstream
projects in my local dev environment. For mmaction2, for example:

Before:

    $ python -m timeit -n1 -r1 "from mmaction.apis import init_recognizer, inference_recognizer"
    1 loop, best of 1: 1.94 sec per loop

After:

    $ python -m timeit -n1 -r1 "from mmaction.apis import init_recognizer, inference_recognizer"
    1 loop, best of 1: 754 msec per loop

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

No

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ X ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ X ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ X ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ X ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ X ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by imabackstabber — APPROVED** [review]
LGTM.

File: mmcv/utils/registry.py [hunk 1]
```diff
@@ -139,9 +139,10 @@ def infer_scope():
         Returns:
             str: The inferred scope name.
         """
-        # inspect.stack() trace where this function is called, the index-2
-        # indicates the frame where `infer_scope()` is called
-        filename = inspect.getmodule(inspect.stack()[2][0]).__name__
```
Comment (imabackstabber) [inline mmcv/utils/registry.py:144]: Suggesting leave a pr tag here as comment, such as:
```
pr #1844 change ``filename = inspect.getmodule(inspect.stack()[2][0]).__name__`` to below to speed up this code segment. 
```

**Review by imabackstabber — CHANGES_REQUESTED** [review]
Test shows it did help.Could you add some comment as suggested? @astahlman

File: mmcv/utils/registry.py (hunk 1 above)
Comment (astahlman) [reply mmcv/utils/registry.py:144]: Sure, I've added a comment with a reference to the PR

**Review by imabackstabber — COMMENTED** [review]
LGTM.

**Review by zhouzaida — APPROVED** [review]

**Comment by astahlman** [comment]
@imabackstabber you've already reviewed this one but Github still has the review in the "Requested changes" state. Mind approving if there are no other issues?

**Comment by zhouzaida** [comment]
> 

@astahlman Thanks for your contribution, I have merged the PR.

**Pull request merged**